core/state_transition.go

```diff
@@ -230,32 +231,39 @@ func (st *StateTransition) TransitionDb() (ret []byte, requiredGas, usedGas *big
 
 	var (
 		evm = st.evm
-		// vm errors do not effect consensus and are therefor
-		// not assigned to err, except for insufficient balance
-		// error.
+		// Vm error, represents the error message returned by
+		// the entire call stack. If vmerr is not empty,
+		// it means that the state revert occurred during the execution.
+		// As ethereum EIP#658 defined, transaction failure means revert happened
+		// due to any operation that can cause the transaction or top-level call.
 		vmerr error
+		// Sub error, represents the outermost call error message,
+		// can only return the error message called at level 0.
+		// Sub error will not been assigned to err, except for insufficient
+		// balance error.
+		suberr error
```

Comment:
I don't think there's a reason for this second suberr, since returning a single error from Create or Call is enough. Below you can check if it's `vm.ErrInsufficientBalance` and `return nil, nil, nil, nil, vmerr` in such a case, and `return ret, requiredGas, st.gasUsed(), vmerr, nil` in the case of the happy path.